src/vm/functions/special.rs

```diff
@@ -0,0 +1,91 @@
+use std::convert::{TryFrom, TryInto};
+use std::cmp;
+
+use vm::types::{Value, PrincipalData, QualifiedContractIdentifier};
+use vm::representations::{SymbolicExpression, SymbolicExpressionType};
+use vm::errors::Error;
+use vm::errors::{CheckErrors, InterpreterError, RuntimeErrorType, InterpreterResult as Result};
+
+use chainstate::stacks::boot::boot_code_id;
+use chainstate::stacks::db::StacksChainState;
+use vm::clarity::ClarityTransactionConnection;
+use vm::database::ClarityDatabase;
+
+fn parse_pox_stacking_result(result: &Value) -> std::result::Result<(PrincipalData, u128, u128), i128> {
+    match result.clone().expect_result() {
+        Ok(res) => {
+            // should have gotten back (ok (tuple (stacker principal) (lock-amount uint) (unlock-burn-height uint)))
+            let tuple_data = res.expect_tuple();
+            let stacker = tuple_data
+                .get("stacker")
+                .expect(&format!("FATAL: no 'stacker'"))
+                .to_owned()
+                .expect_principal();
+
+            let lock_amount = tuple_data
+                .get("lock-amount")
+                .expect(&format!("FATAL: no 'lock-amount'"))
+                .to_owned()
+                .expect_u128();
+
+            let unlock_burn_height = tuple_data
+                .get("unlock-burn-height")
+                .expect(&format!("FATAL: no 'unlock-burn-height'"))
+                .to_owned()
+                .expect_u128();
+
+            Ok((stacker, lock_amount, unlock_burn_height))
+        }
+        Err(e) => {
+            Err(e.expect_i128())
+        }
+    }
+}
+
+/// Handle special cases when calling into the PoX API contract
+fn handle_pox_api_contract_call(db: &mut ClarityDatabase, sender_opt: Option<&PrincipalData>, function_name: &str, value: &Value) -> Result<()> {
+    if function_name == "stack-stx" || function_name == "delegate-stx" {
+        debug!("Handle special-case contract-call to {:?} {} (which returned {:?})", boot_code_id("pox-api"), function_name, value);
+
+        // sender is required
+        let sender = match sender_opt {
+            None => {
+                return Err(RuntimeErrorType::NoSenderInContext.into());
+            }
+            Some(sender) => (*sender).clone()
+        };
+
+        match parse_pox_stacking_result(value) {
+            Ok((stacker, lock_amount, unlock_burn_height)) => {
+                assert_eq!(stacker, sender, "BUG: tx-sender is not contract-call origin!");
```

Comment:
I'm pretty sure this is always true, but I'd appreciate @kantai or @lgalabru's second opinions.  The `(stack-stx)` and `(delegate-stx)` functions return the `tx-sender` as the stacker, since that's whose tokens get locked.  Since this function gets called right after evaluating a `contract-call?`, then it means that the environment's `sender` should always be equal to the `tx-sender` in the contract, right?  i.e. the the `sender` variable comes from the environment that gets used to evaluate the contract-call.

Reply:
If my understanding of the current wiring is correct, `handle_contract_call_special_cases ` would not kicked in if the `stack-stx` call is wrapped inside another contract call. Is it a feature or an issue?

Reply:
Ah, that's not desirable -- the `handle_contract_call_special_cases` needs to run at each `contract-call?` invocation of the PoX API.  Will add test coverage and update.

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1850/commits/40a481e9de28c22912388e07b789f2ce3be44463